### PR TITLE
C++: IR: Fix performance of value-init ranges

### PR DESCRIPTION
**Pull request by jbj**
On a snapshot of Postgres, evaluation of `getNextExplicitlyInitializedElementAfter#fff#antijoin_rhs#1` took forever, preventing the computation of the IR. I haven't been able to reproduce it with a small test case, but the implementation of `getNextExplicitlyInitializedElementAfter` was fragile because it called the inline predicate `ArrayAggregateLiteral.isInitialized`. It also seemed inefficient that `getNextExplicitlyInitializedElementAfter` was computed for many values of its parameters that were never needed by the caller.

This commit replaces `getNextExplicitlyInitializedElementAfter` with a new predicate named `getEndOfValueInitializedRange`, which should have the same behavior but a more efficient implementation. It uses a helper predicate `getNextExplicitlyInitializedElementAfter`, which shares its name with the now-deleted predicate but has behavior that I think matches the name.

I've made this PR against the RC because I worry that the IR will be useless on too many snapshots without this fix.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll [hunk 1]
```diff
@@ -270,7 +270,7 @@ newtype TTranslatedElement =
     not ignoreExpr(initList) and
     isFirstValueInitializedElementInRange(initList, elementIndex) and
     elementCount = 
-      getNextExplicitlyInitializedElementAfter(initList, elementIndex) -
+      getEndOfValueInitializedRange(initList, elementIndex) -
```
Comment (ian-semmle) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll]: Don't you need an `or getNextExplicitlyInitializedElementAfter` too?

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll [hunk 2]
```diff
@@ -322,23 +322,30 @@ newtype TTranslatedElement =
...
-      then result = afterElementIndex + 1
-      else result = getNextExplicitlyInitializedElementAfter(initList, afterElementIndex+1))
-    else
-      result = initList.getType().getUnspecifiedType().(ArrayType).getArraySize() and
-      // required for binding
-      initList.isInitialized(afterElementIndex)
+private int getEndOfValueInitializedRange(ArrayAggregateLiteral initList, int afterElementIndex) {
+  result = getNextExplicitlyInitializedElementAfter(initList, afterElementIndex)
+  or
+  isFirstValueInitializedElementInRange(initList, afterElementIndex) and
```
Comment (ian-semmle) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll]: Shouldn't this be in both halves of the disjunction?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll]: `getNextExplicitlyInitializedElementAfter` calls `isFirstValueInitializedElementInRange`, so adding another call in this predicate would be redundant.

**Comment by jbj** [comment]
It sounds like I ought to add a test that depends on a non-empty `getNextExplicitlyInitializedElementAfter` in order to gain confidence that this change is right. Now that we have designated initializers for arrays, that should be possible.

**Comment by jbj** [comment]
I've added a simple test to exercise the changed predicate. The results are the same before and after this change. The generated IR looks wrong in both cases, though. All the right code is generated, but Block 0 jumps to Block 2 and never lets Block 1 and Block 3 run. Can you explain this, @dave-bartolomeo?

**Comment by jbj** [comment]
Bug filed: https://jira.semmle.com/browse/CPP-260
